[feature] 음원 백그라운드

miller198 opened the pull request:
## #️⃣연관된 이슈

> resolve #123 

## 📝작업 내용 및 코드

1. 음원 백그라운드 재생 - MediaSessionService
2. 픽 정보 화면 및 메인지도에서 음원 재생 중 Configuration Chagne 대응
3. PlayerViewModel 대신 PlayerServiceViewModel 사용 ( PlayerViewModel 는 혹시 모를 롤백 대비 놔두겠습니다 )
4. MediaSessionService 위한 모듈 분리

### MediaPlayerService

MediaSessionService 를 상속받은 객체로,

MediaSessionService 은 대응되는 MediaSession이 활성화될때마다 자동으로 실행되어 Notification을 표시한다.

→ 다른 서비스처럼 startService()나 startForegroundService() 등을 호출하지 않아도 됨

클래스 내부에서 관리할 MediaSession 객체를 생성하고, onGetSession() 메소드에서 리턴하는 MediaSession을 이걸로 지정함으로서 이 MediaSessionService가 관리하는 MediaSession을 지정할 수 있음

여기선 onCreate()에서 서비스가 시작될때 이 MediaSession에 대응되는 Notification 생성을 위해 setMediaNotificationProvider() 호출

setMediaNotificationProvider()의 createNotification() 에서 리턴하는 것은 아래의 MediaNotificationProvider의 MediaNotification 제공함수.

결과적으로 이 MediaSession에 대한 커스텀 Notification을 설정하는 것임

### MediaNotificationProvider

MediaNotification을 생성하고 MediaSession과 연결, 이를 제공하는 클래스

Notification에 필요한 채널 생성(안드로이드 버전 대응), Notification 표시 컨텐츠, Noti에 표시될 커스텀 액션추가, 스타일 및 인텐트 설정을 담당한다.

### CustomMediaSessionCallback

MediaSession.Callback을 상속받아 MediaController로부터 받은 명령에 대한 콜백을 커스터마이징

onConnect 에서 MediaController와 연결 시도할때 동작을 정의한다.

MediaController와 연결되면 이 MediaSession에서 사용할 세션 커맨드들을 추가하고

커맨드에 대응하도록 MediaController에 버튼UI를 추가함(setCustomLayout) → UI에서 사용할 수 있긴한데 컴포즈로는 복잡할듯

여기서 안드로이드 버전별로 세션 커맨드 추가 순서를 뒤집어야했음 (14를 기준으로 설정되는 커맨드 순서가 반대임)

MediaSession을 Build 할때 setCallback 으로 추가되어 프로젝트에  MediaSession

### SessionToken

MediaController 생성에 필요한, MediaSessionService에 대한 토큰

### PlayerCommands

MediaSession 과 MediaComtroller 에서 사용될 커스텀 커맨드들을 정의한 enum

동시에 Notification 설정 시에도 Notification의 addAction을 통해 Noti에 표시될 버튼들이기도함

### MediaControllerProvider

다른 모듈로 MediaController를 전달해 MediaSession 에 대한 제어 방법을 제공하는 클래스

MediaSessionService의 SessionToken을 받아 생성한 MediaController를 callbackFlow로 제공

동시에 MediaSessionService에 있는 ExoPlayer의 AudioSessionId도 callbckFlow로 제공한다 → 비주얼라이징에 사용


## 💬리뷰 요구사항(선택)

픽 정보 화면에서 플레이어 재생 중 메인스레드에서 비동기로 수행하는 작업이 좀 많아서 가끔 튕깁니다.
에뮬레이터 환경에선 자주 튕길수 있을것같아요 너무 튕긴다 싶으면 아깝지만 출시버전에선 그냥 롤백하겠습니다

miller198 committed:
[feature] mediaservice 모듈

miller198 committed:
[feature] 백그라운드 재생 구현

MediaSessionService 이용한 백그라운드 재생 구현,
Notification 표시

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

miller198 committed:
[feature] 메인지도 백그라운드 재생, Config Change

메인지도에 백그라운드 재생구현, 모든 화면 ConfigChange 대응

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

miller198 committed:
[bug] 메인지도 플레이어 버그 수정 중

miller198 committed:
[fix] 메인지도 플레이어 버그 수정

상세화면에서 메인지도로 나왔을때 배너 눌러도 셔플재생 안되고 상세화면에서 듣던 노래만 들리는 문제 해결

miller198 committed:
[chore] player usecase 패키지 분리

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

yuni-ju committed:
[design] pick 정보에서 앨범 영역 높이 지정

miller198 committed:
Merge branch 'feature/#123-background-player' of https://github.com/boostcampwm-2024/and06-musicroad into feature/#123-background-player

miller198 committed:
[fix] 바텀시트 아이템 클릭 stop 누락 수정

yuni-ju commented on settings.gradle.kts:
```diff
@@ -25,3 +25,4 @@ include(":app")
 include(":domain")
 include(":data")
 enableFeaturePreview("TYPESAFE_PROJECT_ACCESSORS")
+include(":mediaservice")
```
include(":mediaservice") 윗 줄로 옮겨주세요 ㅎㅎ

yuni-ju commented on domain/build.gradle.kts:
```diff
@@ -34,10 +36,19 @@ dependencies {
     implementation(libs.androidx.core.ktx)
     implementation(libs.androidx.appcompat)
     implementation(libs.material)
+    implementation(project(":mediaservice"))
```
다른 모듈과 통일되도록 가장 위에 projects로 추가부탁드립니다.

```suggestion
    implementation(projects.mediaservice)
```

yuni-ju commented on app/src/main/java/com/squirtles/musicroad/media/PlayerServiceViewModel.kt:
```diff
@@ -0,0 +1,138 @@
+    private val mediaPlayerUseCase: MediaPlayerUseCase,
+) : ViewModel() {
+
+    private var _playerState: StateFlow<PlayerState> = mediaPlayerUseCase.playerUiStateFlow()
+        .stateIn(
+            scope = viewModelScope,
+            started = SharingStarted.WhileSubscribed(5_000),
+            initialValue = PlayerState()
+        )
+
+    val playerState get() = _playerState
```
혹시 _playerState를 get()으로 한번 더 받는 이유가 있을까요?

miller198 replied:
그러게요 왜저러죠?? ㅋㅋㅋㅋㅋㅋㅋ

yuni-ju commented on domain/src/main/java/com/squirtles/domain/usecase/player/MediaPlayerListenerUseCase.kt:
```diff
@@ -0,0 +1,140 @@
+        coroutineScope.launch {
+            currentPlayerState
+                .onEach { send(it) }
+                .collect { }
```
`.collect()`로 대체할 수 있을 것 같습니다.

miller198 committed:
[fix] Notification 컴팩트 버튼 수 수정

miller198 committed:
[build] 의존성 및 모듈 include 순서 변경

miller198 committed:
[refactor] 리뷰 반영

playerState getter 삭제, collect() 적용

yuni-ju reviewed (approved):
알림창에 뜨는거 신기하네요! 고생하셨습니다~~

miller198 committed:
[refactor] 유저정보페이지 바텀 버튼 위치 수정

meanjoo reviewed (commented):
알림에 뜨는 거 굉장히 신기하네요!! 그런데 노래를 들은 위치에 따라 알림에서 제어할 수 있는 항목이 다른데 저만 그런 걸까요?

|지도 배너를 클릭하여 재생한 후 홈 화면으로 갔을 때|정보 화면에서 노래를 재생한 후 홈 화면으로 갔을 때|
|:---:|:---:|
|![image](https://github.com/user-attachments/assets/4774744f-6e4a-4771-8389-a9ab3897897d)|![image](https://github.com/user-attachments/assets/26b7e4e8-6463-47ab-8319-051b10cb986b)|
|![image](https://github.com/user-attachments/assets/d558b3ce-f290-47cd-8023-d920d3ca9117)|![image](https://github.com/user-attachments/assets/b60bfecd-328b-4e3d-adba-a3900b345f31)|

확장시켰을 때 오른쪽 거는 제어할 수 있는 버튼이 반대인 것 같아요,,!

miller198 committed:
[bug] 앱 종료 버그

음원재생-뮤비-뮤비가로-세로-음원재생 하면 앱 멈춤

miller198 committed:
Revert "[bug] 앱 종료 버그"

This reverts commit 9d8b354c2d8b8ae44ee930fb41d65eb8805b88e1.

miller198 committed:
[bug] 앱 종료 버그

음원재생-뮤비-뮤비가로-세로-음원재생 하면 앱 멈춤

miller198 committed:
[build] version 업그레이드

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

miller198 commented:
@meanjoo 
이거 업데이트해서 확인해봤더니 안드13까진 괜찮은데 안드14부터 저러네요

miller198 committed:
[fix] 버전별 Notification UI 대응

miller198 committed:
[fix] 버그 해결

안드로이드 버전별 Notification UI 대응, 가끔 MediaService 실행 안되어서 플레이어 작동 안되던 문제 해결, 픽 정보에서 다른 노래 재생되는 문제 해결

miller198 committed:
[style] 주석삭제, 로그삭제

miller198 committed:
[refactor] job 변수 삭제

miller198 committed:
[refactor] job 변수 삭제, 주석

miller198 committed:
[refactor] 함수이름변경, 로그삭제

miller198 committed:
[refactor] 버전대응변경, Command 이름 변경

miller198 committed:
Merge branch 'feature/#123-background-player' of https://github.com/boostcampwm-2024/and06-musicroad into feature/#123-background-player

miller198 committed:
[fix] 재생바 및 컨트롤 버그

제어는 되늰데 업데이트안되는 버그

miller198 committed:
[style] debug log 삭제

miller198 committed:
[refactor] 코드 최적화

miller198 committed:
[refactor] 재생속도 명시적 설정, 시각화 애니메이션 속도 변경

miller198 committed:
[refactor] 기본 샘플링 레이트 변경

miller198 committed:
[refactor] MAX 주파수 변경

4000 -> 4400

miller198 committed:
[refactor] 샘플링레이트 수정, 가중치 변경

miller198 committed:
[style] 주석처리 삭제

miller198 committed:
[refactor] 세션 커맨드 추가방식 변경

miller198 committed:
Merge branch 'develop' into feature/#123-background-player

miller198 committed:
[refactor] 세션 커맨드 버전 수정

miller198 committed:
[refactor] 오디오 가중치 수정

meanjoo reviewed (approved):
발표 끝나고 코드 읽어보고 Approve 해야지 하다가 이제야 확인하네요 .. 늦어서 죄송합니다 ㅠ

코드 상에서 궁금한 건 어디에서도 사용되지 않는 함순데 뷰모델에 왜 존재하는 거지? 정도인데 아마 추후 기능 구현을 위한 거라 생각하여 딱히 코멘트를 달지는 않았습니다..!

백그라운드 재생 구현하시느라 고생 많으셨어요!!!👍👍👏
